group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorService.java

```diff
@@ -822,21 +834,60 @@ public CompletableFuture<DeleteGroupsResponseData.DeletableGroupResultCollection
-                    log
-                ));
-
-            futures.add(future);
+            CompletableFuture<DeleteGroupsResponseData.DeletableGroupResultCollection> shareFuture = deleteShareGroups(topicPartition, groupList)
+                .thenCompose(groupErrMap -> {
+                    DeleteGroupsResponseData.DeletableGroupResultCollection collection = new DeleteGroupsResponseData.DeletableGroupResultCollection();
+                    List<String> errGroupIds = new ArrayList<>();
+                    groupErrMap.forEach((groupId, error) -> {
+                        if (error.code() != Errors.NONE.code()) {
+                            log.error("Error deleting share group {} due to error {}", groupId, error);
+                            errGroupIds.add(groupId);
+                            collection.add(
+                                new DeleteGroupsResponseData.DeletableGroupResult()
+                                    .setGroupId(groupId)
+                                    .setErrorCode(error.code())
+                            );
+                        }
+                    });
+
+                    Set<String> groupSet = new HashSet<>(groupList);
+                    // Remove all share group ids which have errored out
+                    // when deleting with persister.
+                    errGroupIds.forEach(groupSet::remove);
+
+                    // If no non-share groupIds or non-error share group ids present
+                    // return.
+                    if (groupSet.isEmpty()) {
+                        return CompletableFuture.completedFuture(collection);
+                    }
+
+                    // Let us invoke the standard procedure of any non-share
+                    // groups or successfully deleted share groups remaining.
+                    List<String> retainedGroupIds = groupSet.stream().toList();
+                    return runtime.scheduleWriteOperation(
+                            "delete-groups",
+                            topicPartition,
+                            Duration.ofMillis(config.offsetCommitTimeoutMs()),
+                            coordinator -> coordinator.deleteGroups(context, retainedGroupIds)
+                        ).thenApply(deletedCollection -> {
+                            deletedCollection.forEach(item -> collection.add(item.duplicate()));
+                            return collection;
+                        })
+                        .exceptionally(exception -> handleOperationException(
+                            "delete-groups",
+                            groupList,
+                            exception,
+                            (error, __) -> {
+                                DeleteGroupsRequest.getErrorResultCollection(retainedGroupIds, error).forEach(item -> collection.add(item.duplicate()));
+                                return collection;
```

Comment:
I don't like the fact that we add to collection in two places (L873 and L881). This is error prone. Could we handle it at the end for both?